common/app/uiComponent/UIComponent.scala

```diff
@@ -0,0 +1,25 @@
+package uiComponent
+
+import model.ApplicationContext
+import play.api.libs.json.JsValue
+import play.api.mvc.Result
+import play.api.mvc.Results._
+import uiComponent.core.JavascriptRenderer
+
+import scala.util.{Failure, Success}
+
+trait UIComponentState {
+  def asJson: JsValue
+}
+
+abstract class UIComponent[-S <: UIComponentState](javascriptFile: String) {
```

Comment:
probably `abstract class UIComponent[-S: UIComponentState](javascriptFile: String)` 

which is sugar for

`abstract class UIComponent[S](javascriptFile: String)(implicit state: UIComponentState[S])` 😃